compiler/circle-opselector/src/OpSelector.cpp

```diff
@@ -0,0 +1,396 @@
+/*
+ * Copyright (c) 2021 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "OpSelector.h"
+
+#include <luci/Import/GraphBuilder.h>
+#include <luci/Import/GraphBuilderContext.h>
+#include <luci/Import/GraphBuilderRegistry.h>
+#include <luci/Import/CircleReader.h>
+#include <luci/Import/Nodes/CircleConst.h>
+
+#include <luci/Profile/CircleNodeID.h>
+#include <luci/Profile/CircleNodeOrigin.h>
+
+#include <loco/IR/Graph.h>
+
+#include <oops/UserExn.h>
+
+#include <cassert>
+#include <iostream>
+#include <map>
+
+#define MAIN_SUBGRAPH 0
+
+// TODO: Re-implement to use just module.
+namespace opselector
+{
+
+void OpSelector::check_connected(std::vector<const luci::CircleNode *> &selected_nodes,
+                                 std::set<uint32_t> &used_output_tensors,
+                                 std::set<uint32_t> &graph_inputs,
+                                 std::set<uint32_t> &graph_outputs)
+{
+  const auto &operators = _reader.operators();
+
+  std::set<uint32_t> selected_input_tensors;
+  std::set<uint32_t> selected_output_tensors;
+
+  std::vector<const circle::OperatorT *> selected_operators;
+
+  // enroll all output nodes.
+  for (auto &op : operators)
+    for (auto output : op.get()->outputs)
+      used_output_tensors.insert(output);
+
+  for (auto input : _reader.inputs()) // graph's input must not have preceding node.
+    used_output_tensors.insert(input);
+
+  // select operators.
+  for (auto cnode : selected_nodes)
+  {
+    uint32_t node_id = luci::get_node_id(cnode);
+    selected_operators.push_back(operators[node_id].get()); // put selected nodes in vector.
+
+    if (cnode->name().find("while") != std::string::npos ||
+        cnode->name().find("if") != std::string::npos) // if has while of if node,
+    {
+      _has_subgraph = true; // A flag indicating whether to copy the subgraph or not,
+    }
+  }
```

Comment:
I'm not sure of this. can there be multiple subgraphs without `IF` or `WHILE` node?

Reply:
No. Without that, no subgraph.
I used it to check if it has `WHILE` or `IF` node.

I didn't know another way to check if a node is `WHILE` or `IF`.
So I checked if the string contains `WHILE` or `IF`.
And if so, I judged that there is a sub-item.